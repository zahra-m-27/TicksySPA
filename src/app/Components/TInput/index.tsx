import {Tag} from 'antd';
import styles from './styles.module.scss';
import React, {InputHTMLAttributes, useEffect, useState} from 'react';
import ClassNames from '../../Utilities/ClassNames';

interface Item {
  value: any;
  label: string;
}

interface Props extends InputHTMLAttributes<HTMLInputElement> {
  icon?: string;
  label?: string;
  regex?: RegExp;
  tags?: string[];
  content?: string;
  minLines?: number;
  hasError?: boolean;
  className?: string;
  isNumeric?: boolean;
  onEnter?: () => void;
  iconClassName?: string;
  inputClassName?: string;
  labelClassName?: string;
  autoCompleteData?: Item[];
  autoCompleteClassName?: string;
  inputContainerClassName?: string;
  autoCompleteItemClassName?: string;
  onTagClose?: (index: number) => void;
  onAutoCompleteSelect?: (value: any) => void;
  onChangeText?: (text: string, HasError: boolean) => void;
}

export default function TInput({
  icon,
  tags,
  label,
  regex,
  content,
  onEnter,
  minLines,
  className,
  onTagClose,
  onChangeText,
  iconClassName,
  inputClassName,
  labelClassName,
  autoCompleteData,
  hasError = false,
  isNumeric = false,
  onAutoCompleteSelect,
  autoCompleteClassName,
  autoCompleteItemClassName,
  inputContainerClassName,
  ...props
}: Props) {
  const tagColors = ['#9fa8b1'];
  const [HasError, setHasError] = useState(false);
  const [Content, setContent] = useState(content ?? '');

  useEffect(() => {
    setContent(content ?? '');
  }, [content]);

  useEffect(() => {
    setHasError(hasError);
  }, [hasError]);

  const onChange = (event: React.ChangeEvent<any>) => {
    if (isNumeric && !/^\d*$/.test(event.target.value)) {
      setContent(event.target.value);
      setTimeout(() => {
        setContent('');
        setHasError(true);
      }, 150);
      return;
    }

    if (regex && !regex.test(event.target.value)) {
      setHasError(true);
    } else if (HasError) {
      setHasError(false);
    }

    setContent(event.target.value);
    onChangeText && onChangeText(event.target.value, HasError);
  };

  let input_style = ClassNames(styles.input_class, inputClassName);
  if (HasError) {
    input_style = ClassNames(styles.input_class, inputClassName);
  }

  let input_container_style = ClassNames(
    styles.input_container,
    inputContainerClassName
  );
  if (HasError) {
    input_container_style = ClassNames(
      styles.input_container,
      inputContainerClassName,
      styles.failed_regex_focused_input_container
    );
  }

  const label_style = ClassNames(styles.label, labelClassName);
  const container_style = ClassNames(styles.container, className);

  return (
    <div className={container_style}>
      <label className={label_style}>{label}</label>
      <div className={input_container_style} data-testid="input-container">
        {icon && (
          <img
            src={icon}
            className={[styles.icon, iconClassName].join(' ')}
            alt="icon"
          />
        )}
        {minLines ? (
          <textarea
            dir="rtl"
            rows={minLines}
            value={Content}
            onChange={onChange as any}
            className={input_style}
            {...(props as any)}
          />
        ) : (
          <input
            dir="rtl"
            value={Content}
            onChange={onChange}
            className={input_style}
            onKeyDown={(e) => e.key === 'Enter' && onEnter && onEnter()}
            {...props}
          />
        )}
      </div>

      <div
        className={ClassNames(
          styles.auto_complete_items,
          autoCompleteClassName
        )}
        data-active={!!autoCompleteData && !!autoCompleteData.length}>
        {autoCompleteData &&
          autoCompleteData.map((item) => (
            <div
              key={`${item.value}`}
              className={ClassNames(
                styles.auto_complete_item,
                autoCompleteItemClassName
              )}
              onClick={() =>
                onAutoCompleteSelect && onAutoCompleteSelect(item.value)
              }>
              <p>{item.label}</p>
            </div>
          ))}
      </div>

      {tags && tags.length > 0 && (
        <div className={styles.tags_container}>
          {tags.map((tag, i) => (
            <Tag
              key={tag}
              closable
              color={tagColors[i % tagColors.length]}
              onClose={() => onTagClose && onTagClose(i)}>
              #{tag}
            </Tag>
          ))}
        </div>
      )}
    </div>
  );
}
