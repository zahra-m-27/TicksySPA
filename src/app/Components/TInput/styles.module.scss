.container {
  width: 100%;
  display: flex;
  align-items: flex-end;
  flex-direction: column;
  font-size: clamp(14px, 1.1vw, 18px);
}

.input_class {
  width: 100%;
  padding: 10px;
  border-radius: 5px;
  background: #dadde0;
  border: 1px solid #e3e3e3;
}

.label {
  color: #333333;
  margin-bottom: 10px;
}

.failed_regex_focused_input_container {
  border-color: #e62e00;
  box-shadow: 0 0 8px 0 #e62e00;
}

.tags_container {
  width: 100%;
  display: flex;
  margin: 0 10px;
  padding: 10px 0;
  overflow-x: auto;
}
