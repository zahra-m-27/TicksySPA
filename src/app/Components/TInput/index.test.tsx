import React from 'react';
import {render, waitFor} from '@testing-library/react';
import TInput from './index';
import ReactDOM from 'react-dom';
import user from '@testing-library/user-event';
import renderer from 'react-test-renderer';

const onChange = jest.fn();

const sample = (
  <TInput
    name="email"
    label="Test Label"
    content="Test Content"
    onChangeText={onChange}
  />
);
const sample2 = (
  <TInput
    isNumeric
    name="number"
    label="Test Label"
    content="123456"
    onChangeText={onChange}
  />
);
const sample3 = (
  <TInput
    isNumeric
    name="number"
    regex={/\d\d/}
    content="123456"
    label="Test Label"
    onChangeText={onChange}
  />
);
const sample4 = (
  <TInput name="tags" label="Tags" tags={['tag1']} onChangeText={onChange} />
);

test('render in ReactDOM', () => {
  const div = document.createElement('div');
  ReactDOM.render(sample, div);
});

test('renders label', () => {
  const {getByText} = render(sample);

  getByText(/Test Label/i);
});

test('renders input value', () => {
  const {getByDisplayValue} = render(sample);

  getByDisplayValue(/Test Content/i);
});

test('pass input parameter', () => {
  const {getByDisplayValue} = render(sample);

  const inputElement = getByDisplayValue(/Test Content/i);
  expect(
    inputElement.attributes.getNamedItem('name')?.value === 'email'
  ).toBeTruthy();
});

it('input onChange', () => {
  const {getByDisplayValue} = render(sample);
  const inputElement = getByDisplayValue(/Test Content/i);
  user.clear(inputElement);
  expect(onChange).toBeCalledWith('', expect.anything());
  user.type(inputElement, 'Test Typing');
  expect(onChange).toBeCalledWith('Test Typing', expect.anything());
});

it('number input onChange', async () => {
  const {getByDisplayValue, getByTestId} = render(sample2);
  const inputElement = getByDisplayValue(/123456/i);
  user.clear(inputElement);
  expect(onChange).toBeCalledWith('', expect.anything());
  user.type(inputElement, '1234');
  expect(onChange).toBeCalledWith('1234', expect.anything());
  user.clear(inputElement);
  expect(onChange).toBeCalledWith('', expect.anything());
  user.type(inputElement, 'R');
  await waitFor(() =>
    expect(getByTestId('input-container').className).toContain(
      'failed_regex_focused_input_container'
    )
  );
  user.clear(inputElement);
  user.type(inputElement, '124');
  expect(onChange).toBeCalledWith('124', expect.anything());
});

it('regex check', () => {
  const {getByDisplayValue} = render(sample3);
  const inputElement = getByDisplayValue(/123456/i);
  user.clear(inputElement);
  expect(onChange).toBeCalledWith('', expect.anything());
  user.type(inputElement, '1234');
});

test('renders tags', () => {
  const {getByText} = render(sample4);

  const tag = getByText(/tag1/i);

  const closeTag = tag.querySelector('.anticon-close');

  if (closeTag) {
    user.click(closeTag);
  }
});

it('snapshot Input', () => {
  const snapshot = renderer.create(sample).toJSON();
  expect(snapshot).toMatchSnapshot();
});
