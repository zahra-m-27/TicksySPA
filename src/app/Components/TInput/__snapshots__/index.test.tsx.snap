// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Input 1`] = `
<div
  className="container"
>
  <label
    className="label"
  >
    Test Label
  </label>
  <div
    className="input_container"
    data-testid="input-container"
  >
    <input
      className="input_class"
      dir="auto"
      name="email"
      onChange={[Function]}
      onKeyDown={[Function]}
      value="Test Content"
    />
  </div>
</div>
`;
